sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -911,6 +911,11 @@ object CollapseRepartition extends Rule[LogicalPlan] {
     // we can remove the child.
     case r @ RepartitionByExpression(_, child: RepartitionOperation, _) =>
       r.copy(child = child.child)
+
+    // Case 3: When a RebalancePartitions has a child of Repartition or RepartitionByExpression
+    // we can remove the child.
```

Comment:
No, we can't do this. `RebalancePartitions` is a best-effort and does not guarantee partitioning. We intentionally do not let `RebalancePartitions` not extend `RepartitionOperation`, to avoid wrong optimizations.

Please revert this.

Reply:
See https://github.com/apache/spark/pull/32932#discussion_r656249153

Reply:
OK

Reply:
Reverted at: https://github.com/apache/spark/commit/108635af1708173a72bec0e36bf3f2cea5b088c4

Reply:
Oh, got it. Thank you for reverting, @cloud-fan and @wangyum .